Model/src/main/java/org/gusdb/wdk/model/fix/table/edaanalysis/plugins/VDIMigrationPlugin.java

```diff
@@ -1,89 +1,153 @@
+
+    // Validate required args.
+    if (!args.containsKey("--tinyDb")) {
+      throw new IllegalArgumentException("Missing required flag --tinyDb");
+    }
+    if (!args.containsKey(("--schema"))) {
+      throw new IllegalArgumentException("Missing required argument --schema");
+    }
+
+    final String schema = args.get("--schema");
+    setEntityIdRetriever(new VDIEntityIdRetriever(wdkModel.getAppDb().getDataSource(), schema));
+
+    final File tinyDbFile = new File(args.get("--tinyDb"));
+    readVdiMappingFile(tinyDbFile);
+
+    // Default to dryrun to avoid incidental migrations when testing.
+    _writeToDb = Boolean.parseBoolean(args.getOrDefault("-write", "false"));
+    _wdkModel = wdkModel;
+  }
+
+  // Visible for testing.
+  void setEntityIdRetriever(VDIEntityIdRetriever entityIdRetriever) {
+    _vdiEntityIdRetriever = entityIdRetriever;
+  }
+
+  // Visible for testing
+  void readVdiMappingFile(File mappingFile) {
+    _legacyIdToVdiId = readLegacyStudyIdToVdiId(mappingFile);
+  }
+
   @Override
   public TableRowInterfaces.RowResult<AnalysisRow> processRecord(AnalysisRow nextRow) throws Exception {
     final String legacyDatasetId = nextRow.getDatasetId();
+
     if (!legacyDatasetId.startsWith(UD_DATASET_ID_PREFIX)) {
-      return new TableRowInterfaces.RowResult<>(nextRow)
-          .setShouldWrite(false);
+      return new TableRowInterfaces.RowResult<>(nextRow).setShouldWrite(false);
     }
 
     final String legacyUdId = legacyDatasetId.replace(UD_DATASET_ID_PREFIX, "");
-    final String vdiId = legacyIdToVdiId.get(legacyUdId);
+    final String vdiId = _legacyIdToVdiId.get(legacyUdId);
 
     if (vdiId == null) {
       LOG.warn("Unable to find legacy ID " + legacyUdId + " in the tinydb file.");
       missingFromVdiCount.incrementAndGet();
-      return new TableRowInterfaces.RowResult<>(nextRow);
+      return new TableRowInterfaces.RowResult<>(nextRow)
+          .setShouldWrite(false);
     }
 
     // Append UD prefix to VDI ID. The prefix is prepended in the view that maps stable VDI IDs to the unstable study
     // ID, which is the currency of EDA.
     final String vdiDatasetId = UD_DATASET_ID_PREFIX + vdiId;
-    nextRow.setDatasetId(vdiDatasetId);
+    final Optional<String> vdiEntityId = _vdiEntityIdRetriever.queryEntityId(vdiDatasetId);
+    if (vdiEntityId.isEmpty()) {
+      LOG.warn("Unable to find entity ID in appdb for VDI dataset ID: " + vdiDatasetId);
+      return new TableRowInterfaces.RowResult<>(nextRow)
+          .setShouldWrite(false);
+    }
+
+    LOG.info("Analysis descriptor before migration: " + nextRow.getDescriptor());
+    String descriptor = nextRow.getDescriptor().toString();
+
+    // Find all variable IDs.
+    final Set<String> legacyVariableIds = VAR_ID_PATTERN.matcher(descriptor).results()
+        .map(match -> match.group(1))
+        .collect(Collectors.toSet());
+
+    final String entityId = ENTITY_ID_PATTERN.matcher(descriptor).results()
```

Comment:
Are we sure there's only one here?  We could easily have a config with >1 variable spec, each of which might contain a different entity ID, especially across different visualizations.

Reply:
There's only one entity allowed in ISA simple studies